### PR TITLE
Add benchmark script

### LINKED ISSUE #11: PCC: benchmark storage performance
related to / blocked by #10:

For each of the types of reads and writes for the PCC, we will want to make sure our storage choice is performant:

- given the (large) quantity of data already in storage
- given the volume of read/write requests expected

Can think of this from perspective of:
- Trusted Checksum Repo
- Object Inventory
- Object Provenance

and/or from

- Audit process
  - Inventory process?
- Archive process
- Ingest process

### PR DESCRIPTION
**Pull request by tallenaz**
Closes #11

### CODE REVIEW

File: lib/benchmarking/postgres_testing.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  test_data_lines = File.readlines(test_data_file)
+else
+  logger.warn "I can't find your test data!"
+end
+
+# read the test file into a data structure (array of hashes):
+# { druid: 'bb07ny0600',
+#   version: 'v0002',
+#   version_time: 1490296357,
+#   endpoint: 'sf6-sdr:/vol/sdr_services_10' }
```
Comment (ndushay) [inline lib/benchmarking/postgres_testing.rb]: i think this comment may have become superfluous?

File: lib/benchmarking/postgres_testing.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  moab[:size] = 250
+  moab[:current_version] = line[1]
+  moab_array << moab
+end
+
+# foreach thing, store it into the db with active record
+# PreservedObject.create!(
+#   druid: @id, size: moab_size(@id), current_version: Stanford::StorageServices.current_version(@id)
+# )
+# how long does it take?
```
Comment (ndushay) [inline lib/benchmarking/postgres_testing.rb]: similarly, this comment may be superfluous now? or keep line 57 only?

**Comment by tallenaz** [comment]
@ndushay I have the script set up with `Benchmark.realtime` now. Sample log output:

```
I, [2017-09-12T11:11:58.305323 #11152]  INFO -- : It took 4047.9006802621298 seconds to populate the database
I, [2017-09-12T11:11:58.305493 #11152]  INFO -- : There are 886416 records in the database
I, [2017-09-12T11:11:58.305527 #11152]  INFO -- : The process loaded 218.98165741126792 records per second
```

**Comment by tallenaz** [comment]
I'm looking into how to get coveralls to ignore this script, for the sake of coverage.

File: lib/benchmarking/postgres_testing.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  moab = {}
+  line = line.split(' ')
+  moab[:id] = line[0]
+  moab[:size] = 250
+  moab[:current_version] = line[1]
+  moab_array << moab
+end
+
+time_to_populate_db = Benchmark.realtime do
+  moab_array.each do |moab|
```
Comment (ndushay) [inline lib/benchmarking/postgres_testing.rb]: interesting.  I was thinking you might have benchmark track each create and collate the stats for you ... but whatever.

**Comment by ndushay** [comment]
LGTM;  👍  to merge.  

maybe in future we'll consider benchmarking each write.  Here's some code I wrote for analyzing timings of WAS robot steps:

(NOTE esp: the loop using `Benchmark.benchmark do |x|` and `x.report` stuff.

```
class CdxBigMergeTimings

  def self.benchmark(cmd, count)
    p "  timing: #{cmd}"
    Benchmark.benchmark(CAPTION, 10, FORMAT, ">avg resp:") do |x|
      total_resp = Benchmark::Tms.new
      count.times.each { |i|
        resp_t = x.report("iter #{i}") do
          unless system("#{cmd}_#{count}")
            fail "Error with command: #{cmd}\n" + $CHILD_STATUS.to_s
          end
        end
        total_resp += resp_t
      }
      p '-----------------'
      [total_resp/count]
    end
    nil
  end

  indices_dir = '/web-archiving-stacks/data/indices'
  cdx_working_dir = "#{indices_dir}/cdx_working"
  tmp_dir = '/web-archiving-stacks/data/tmp/'
  cmd_env_vars = "TMPDIR=#{tmp_dir} LC_ALL=C"
  druid_merged_file = "#{cdx_working_dir}/pf139tj8228_one_step_u_merged.cdx_3"
  main_cdx_file = "#{indices_dir}/cdx/index.cdx"
  working_sorted_cdx_file = "#{cdx_working_dir}/pf139tj8228_sorted_index.cdx"

  count = 3

  p '================================================'
  p '==== sort -m -u ===='
  p '================================================'
  cmd = "#{cmd_env_vars} /usr/bin/time -v sort -m -u #{druid_merged_file} #{main_cdx_file} > #{working_sorted_cdx_file}"
  CdxBigMergeTimings.benchmark(cmd, 2)
  p '================================================'
  p

  p '================================================'
  p '==== sort -m -u -S 4M ===='
  p '================================================'
  cmd = "#{cmd_env_vars} /usr/bin/time -v sort -m -u -S 4M #{druid_merged_file} #{main_cdx_file} > #{working_sorted_cdx_file}"
  CdxBigMergeTimings.benchmark(cmd, count)
  p '================================================'
  p

  p '================================================'
  p '==== sort -m -u -S 2G ===='
  p '================================================'
  cmd = "#{cmd_env_vars} /usr/bin/time -v sort -m -u -S 2G #{druid_merged_file} #{main_cdx_file} > #{working_sorted_cdx_file}"
  CdxBigMergeTimings.benchmark(cmd, count)
  p '================================================'
  p

  p '================================================'
  p '==== sort -m -u -S 50% ===='
  p '================================================'
  cmd = "#{cmd_env_vars} /usr/bin/time -v sort -m -u -S 50% #{druid_merged_file} #{main_cdx_file} > #{working_sorted_cdx_file}"
  CdxBigMergeTimings.benchmark(cmd, count)
  p '================================================'
  p
```

**Comment by ndushay** [comment]
Dunno if you wanna make a ticket to do that or ignore it for now or what ...

**Pull request merged**